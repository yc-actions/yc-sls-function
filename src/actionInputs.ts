export type ActionInputs = {
    folderId: string
    functionName: string
    runtime: string
    entrypoint: string
    memory: number
    include: string[]
    excludePattern: string[]
    sourceRoot: string
    executionTimeout: number
    environment: string[]
    serviceAccount: string
    serviceAccountName: string
    bucket: string
    description: string
    secrets: string[]
    networkId: string
    tags: string[]
    logsDisabled: boolean
    logsGroupId: string
    logLevel: number

    async: boolean
    asyncSaId: string
    asyncSaName: string
    asyncRetriesCount: number
    asyncSuccessYmqArn: string
    asyncSuccessSaId: string
    asyncFailureYmqArn: string
    asyncFailureSaId: string
    asyncSuccessSaName: string
    asyncFailureSaName: string
}
