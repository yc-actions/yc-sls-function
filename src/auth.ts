import { info } from '@actions/core'
import axios from 'axios'

export async function exchangeToken(token: string, saId: string): Promise<string> {
    info(`Exchanging token for service account ${saId}`)
    const res = await axios.post(
        'https://auth.yandex.cloud/oauth/token',
        {
            grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
            requested_token_type: 'urn:ietf:params:oauth:token-type:access_token',
            audience: saId,
            subject_token: token,
            subject_token_type: 'urn:ietf:params:oauth:token-type:id_token'
        },
        {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            }
        }
    )
    if (res.status !== 200) {
        throw new Error(`Failed to exchange token: ${res.status} ${res.statusText}`)
    }
    if (!res.data.access_token) {
        throw new Error(`Failed to exchange token: ${res.data.error} ${res.data.error_description}`)
    }
    info(`Token exchanged successfully`)
    return res.data.access_token
}
