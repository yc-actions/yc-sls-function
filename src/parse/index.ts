export { parseMounts } from './mounts'
export { parseIgnoreGlobPatterns } from './globPatterns'
export { parseEnvironmentVariables } from './environmentVariables'
export { parseLockboxVariables, type Secret } from './lockboxVariables'
export { parseLogLevel } from './logLevel'
export { parseMemory, GB, KB, MB } from './memory'
export { parseServiceAccountJsonFile, type ServiceAccountJsonFileContents } from './saJson'
