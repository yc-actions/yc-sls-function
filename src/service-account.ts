import { serviceClients, Session } from '@yandex-cloud/nodejs-sdk'
import { ListServiceAccountsRequest } from '@yandex-cloud/nodejs-sdk/dist/generated/yandex/cloud/iam/v1/service_account_service'

export async function resolveServiceAccountId(
    session: Session,
    folderId: string,
    serviceAccountId: string,
    serviceAccountName: string
): Promise<string | undefined> {
    const client = session.client(serviceClients.ServiceAccountServiceClient)
    if (serviceAccountId) {
        return serviceAccountId
    }
    if (serviceAccountName) {
        const { serviceAccounts } = await client.list(
            ListServiceAccountsRequest.fromPartial({
                folderId,
                filter: `name = "${serviceAccountName}"`
            })
        )

        if (serviceAccounts.length === 0) {
            throw new Error(`Service account with name ${serviceAccountName} not found`)
        }
        return serviceAccounts[0].id
    }
    return undefined
}
