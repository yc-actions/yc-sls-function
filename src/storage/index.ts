import { IamTokenService } from '@yandex-cloud/nodejs-sdk/dist/token-service/iam-token-service'
import axios from 'axios'
import { IStorageObject, StorageObject } from './storage-object'
import { SessionConfig } from '@yandex-cloud/nodejs-sdk/dist/types'

export interface StorageService {
    getObject(bucketName: string, objectName: string): Promise<IStorageObject>

    putObject(object: object): Promise<void>
}

export class StorageServiceImpl implements StorageService {
    static __endpointId = 'storage'
    private readonly _address: string = 'storage.yandexcloud.net:443'
    private readonly _tokenCreator: () => Promise<string>
    private $method_definitions: unknown

    constructor(sessionConfig: SessionConfig) {
        if ('serviceAccountJson' in sessionConfig) {
            const ts = new IamTokenService(sessionConfig.serviceAccountJson)
            this._tokenCreator = async () => ts.getToken()
        } else if ('iamToken' in sessionConfig) {
            this._tokenCreator = async () => sessionConfig.iamToken
        } else throw new Error('IAMToken not implemented.')

        this.$method_definitions = {}
    }

    async getObject(bucketName: string, objectName: string): Promise<StorageObject> {
        const token = await this._tokenCreator()
        const res = await axios.get(this.#_url(bucketName, objectName), {
            headers: {
                'X-YaCloud-SubjectToken': token
            }
        })
        const buf = await res.data()
        return StorageObject.fromBuffer(bucketName, objectName, buf)
    }

    async putObject({ bucketName, bufferPromise, objectName }: IStorageObject): Promise<void> {
        const token = await this._tokenCreator()
        const buffer = await bufferPromise
        await axios.put(this.#_url(bucketName, objectName), buffer, {
            headers: {
                'X-YaCloud-SubjectToken': token
            },
            maxBodyLength: 128 * 1024 * 1024
        })
    }

    #_url(bucketName: string, objectName: string): string {
        return `https://${this._address}/${bucketName}/${objectName}`
    }
}
