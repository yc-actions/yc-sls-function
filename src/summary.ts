import { summary } from '@actions/core'

export async function writeSummary({
    functionName,
    functionId,
    versionId,
    bucket,
    bucketObjectName,
    errorMessage,
    folderId
}: {
    functionName?: string
    functionId?: string
    versionId?: string
    bucket?: string
    bucketObjectName?: string
    errorMessage?: string
    folderId?: string
}) {
    const items: string[] = []
    if (functionName) items.push(`Function Name: ${functionName}`)
    if (functionId && folderId) {
        const url = `https://console.yandex.cloud/folders/${folderId}/functions/functions/${functionId}/overview`
        items.push(`Function ID: <a href="${url}">${functionId}</a>`)
    }
    if (versionId) items.push(`Version ID: ${versionId}`)
    if (bucket) items.push(`Bucket: ${bucket}`)
    if (bucketObjectName) items.push(`Bucket Object: ${bucketObjectName}`)
    if (errorMessage) {
        items.push(`❌ Error: ${errorMessage}`)
    } else {
        items.push('✅ Success')
    }
    if (items.length === 0) return
    await summary.addHeading('Yandex Cloud Function Deployment Summary', 2).addList(items).write()
}
