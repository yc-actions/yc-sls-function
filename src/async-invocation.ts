import { ActionInputs } from './actionInputs'
import {
    AsyncInvocationConfig,
    AsyncInvocationConfig_ResponseTarget
} from '@yandex-cloud/nodejs-sdk/dist/generated/yandex/cloud/serverless/functions/v1/function'
import { resolveServiceAccountId } from './service-account'
import { Session } from '@yandex-cloud/nodejs-sdk'

export function isAsync(actionInputs: ActionInputs): boolean {
    return actionInputs.async
}

export function validateAsync(actionInputs: ActionInputs): boolean {
    if (!isAsync(actionInputs)) {
        return true
    }

    // either asyncSuccessSaId or asyncSuccessSaName must be set if asyncSuccessYmqArn is set
    if (actionInputs.asyncSuccessYmqArn) {
        if (!actionInputs.asyncSuccessSaId && !actionInputs.asyncSuccessSaName) {
            throw new Error(
                'Either async-success-sa-id or async-success-sa-name must be set if async-success-ymq-arn is set'
            )
        }
        // but not both
        if (actionInputs.asyncSuccessSaId && actionInputs.asyncSuccessSaName) {
            throw new Error('Either async-success-sa-id or async-success-sa-name must be set, but not both')
        }
    }
    // either asyncFailureSaId or asyncFailureSaName must be set if asyncFailureYmqArn is set
    if (actionInputs.asyncFailureYmqArn) {
        if (!actionInputs.asyncFailureSaId && !actionInputs.asyncFailureSaName) {
            throw new Error(
                'Either async-failure-sa-id or async-failure-sa-name must be set if async-failure-ymq-arn is set'
            )
        }
        // but not both
        if (actionInputs.asyncFailureSaId && actionInputs.asyncFailureSaName) {
            throw new Error('Either async-failure-sa-id or async-failure-sa-name must be set, but not both')
        }
    }

    return true
}

export async function createAsyncInvocationConfig(
    session: Session,
    actionInputs: ActionInputs
): Promise<AsyncInvocationConfig | undefined> {
    if (!isAsync(actionInputs)) {
        return undefined
    }

    let successTarget: AsyncInvocationConfig_ResponseTarget
    let failureTarget: AsyncInvocationConfig_ResponseTarget

    if (actionInputs.asyncSuccessYmqArn) {
        let serviceAccountId: string | undefined = actionInputs.asyncSuccessSaId
        if (!serviceAccountId && actionInputs.asyncSuccessSaName) {
            serviceAccountId = await resolveServiceAccountId(
                session,
                actionInputs.folderId,
                actionInputs.asyncSuccessSaId,
                actionInputs.asyncSuccessSaName
            )
        }

        successTarget = AsyncInvocationConfig_ResponseTarget.fromPartial({
            ymqTarget: {
                queueArn: actionInputs.asyncSuccessYmqArn,
                serviceAccountId
            }
        })
    } else {
        successTarget = AsyncInvocationConfig_ResponseTarget.fromPartial({
            emptyTarget: {}
        })
    }

    if (actionInputs.asyncFailureYmqArn) {
        let serviceAccountId: string | undefined = actionInputs.asyncFailureSaId
        if (!serviceAccountId && actionInputs.asyncFailureSaName) {
            serviceAccountId = await resolveServiceAccountId(
                session,
                actionInputs.folderId,
                actionInputs.asyncFailureSaId,
                actionInputs.asyncFailureSaName
            )
        }

        failureTarget = AsyncInvocationConfig_ResponseTarget.fromPartial({
            ymqTarget: {
                queueArn: actionInputs.asyncFailureYmqArn,
                serviceAccountId
            }
        })
    } else {
        failureTarget = AsyncInvocationConfig_ResponseTarget.fromPartial({
            emptyTarget: {}
        })
    }

    let serviceAccountId = await resolveServiceAccountId(
        session,
        actionInputs.folderId,
        actionInputs.asyncSaId,
        actionInputs.asyncSaName
    )
    if (serviceAccountId === undefined) {
        serviceAccountId = await resolveServiceAccountId(
            session,
            actionInputs.folderId,
            actionInputs.serviceAccount,
            actionInputs.serviceAccountName
        )
    }

    return AsyncInvocationConfig.fromPartial({
        retriesCount: actionInputs.asyncRetriesCount,
        successTarget,
        failureTarget,
        serviceAccountId
    })
}
