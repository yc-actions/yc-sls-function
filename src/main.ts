import * as core from '@actions/core';
import * as github from '@actions/github';
import archiver from 'archiver';
import * as streamBuffers from 'stream-buffers';
import minimatch from 'minimatch';

import {StorageObject} from '@nikolay.matrosov/yc-ts-sdk/lib/src/storage/v1beta/storageObject';
import {StorageService} from '@nikolay.matrosov/yc-ts-sdk/lib/src/storage/v1beta';
import {Session} from '@nikolay.matrosov/yc-ts-sdk';
import {completion, getMetadata} from '@nikolay.matrosov/yc-ts-sdk/lib/src/operation';
import {FunctionService} from '@nikolay.matrosov/yc-ts-sdk/lib/api/serverless/functions/v1';
import {fromServiceAccountJsonFile} from '@nikolay.matrosov/yc-ts-sdk/lib/src/TokenService/iamTokenService';
import {
  CreateFunctionMetadata,
  CreateFunctionRequest,
  CreateFunctionVersionMetadata,
  CreateFunctionVersionRequest,
  FunctionServiceService,
  ListFunctionsRequest,
} from '@nikolay.matrosov/yc-ts-sdk/lib/generated/yandex/cloud/serverless/functions/v1/function_service';
import {Client} from 'nice-grpc';
import {Operation} from '@nikolay.matrosov/yc-ts-sdk/lib/generated/yandex/cloud/operation/operation';
import {KB, parseMemory} from './memory';
import {Package} from '@nikolay.matrosov/yc-ts-sdk/lib/generated/yandex/cloud/serverless/functions/v1/function';
import * as fs from 'fs';

type ActionInputs = {
  folderId: string;
  functionName: string;
  runtime: string;
  entrypoint: string;
  memory: number;
  include: string[];
  excludePattern: string[];
  executionTimeout: number;
  environment: string[];
  serviceAccount: string;
  bucket: string;
  description: string;
};

async function uploadToS3(bucket: string, functionId: string, session: Session, fileContents: Buffer): Promise<string> {
  const {GITHUB_SHA} = process.env;

  if (!GITHUB_SHA) {
    core.setFailed('Missing GITHUB_SHA');
    throw new Error('Missing GITHUB_SHA');
  }

  //setting object name
  const bucketObjectName = `${functionId}/${GITHUB_SHA}.zip`;
  core.info(`Upload to bucket: "${bucket}/${bucketObjectName}"`);

  const storageService = StorageService(session);

  const storageObject = StorageObject.fromBuffer(bucket, bucketObjectName, fileContents);
  await storageService.putObject(storageObject);
  return bucketObjectName;
}

async function getOrCreateFunctionId(
  session: Session,
  functionService: Client<typeof FunctionServiceService, {}>,
  {folderId, functionName}: ActionInputs,
): Promise<string> {
  core.startGroup('Find function id');
  const res = await functionService.list(
    ListFunctionsRequest.fromPartial({
      folderId,
      filter: `name = '${functionName}'`,
    }),
  );
  let functionId: string;
  // If there is a function with the provided name in given folder, then return its id
  if (res.functions.length) {
    functionId = res.functions[0].id;
    core.info(`'There is the function named '${functionName}' in the folder already. Its id is '${functionId}'`);
  } else {
    // Otherwise create new a function and return its id.
    const repo = github.context.repo;

    const op = await functionService.create(
      CreateFunctionRequest.fromPartial({
        folderId,
        name: functionName,
        description: `Created from ${repo.owner}/${repo.repo}`,
      }),
    );
    await completion(op, session);
    functionId = (getMetadata(op) as CreateFunctionMetadata).functionId;
    core.info(`There was no function named '${functionName}' in the folder. So it was created. Id is '${functionId}'`);
  }
  core.setOutput('function-id', functionId);
  core.endGroup();
  return functionId;
}

async function run(): Promise<void> {
  core.setCommandEcho(true);

  try {
    const ycSaJsonCredentials = core.getInput('yc-sa-json-credentials', {
      required: true,
    });
    core.setSecret(ycSaJsonCredentials);

    const serviceAccountJson = fromServiceAccountJsonFile(JSON.parse(ycSaJsonCredentials));

    const inputs: ActionInputs = {
      folderId: core.getInput('folder-id', {required: true}),
      functionName: core.getInput('function-name', {required: true}),
      runtime: core.getInput('runtime', {required: true}),
      entrypoint: core.getInput('entrypoint', {required: true}),
      memory: parseMemory(core.getInput('revision-memory', {required: false}) || '128Mb'),
      include: core.getMultilineInput('include', {required: false}),
      excludePattern: core.getMultilineInput('exclude', {required: false}),
      executionTimeout: parseInt(core.getInput('execution-timeout', {required: false}) || '5', 10),
      environment: core.getMultilineInput('environment', {required: false}),
      serviceAccount: core.getInput('service-account', {required: false}),
      bucket: core.getInput('bucket', {required: false}),
      description: core.getInput('description', {required: false}),
    };

    core.info('Function inputs set');

    const fileContents = await zipSources(inputs);

    core.info(`Buffer size: ${Buffer.byteLength(fileContents)}b`);

    // Initialize SDK with your token
    const session = new Session({serviceAccountJson});
    const functionService = FunctionService(session);

    const functionId = await getOrCreateFunctionId(session, functionService, inputs);
    let bucketObjectName = '';
    if (inputs.bucket) {
      bucketObjectName = await uploadToS3(inputs.bucket, functionId, session, fileContents);
    }

    await createFunctionVersion(session, functionService, functionId, fileContents, bucketObjectName, inputs);

    core.setOutput('time', new Date().toTimeString());
  } catch (error) {
    if (error instanceof Error) {
      core.setFailed(error.message);
    }
  }
}

function handleOperationError(operation: Operation): void {
  if (operation.error) {
    const details = operation.error?.details;
    if (details) {
      throw Error(`${operation.error.code}: ${operation.error.message} (${details.join(', ')})`);
    }

    throw Error(`${operation.error.code}: ${operation.error.message}`);
  }
}

async function createFunctionVersion(
  session: Session,
  functionService: Client<typeof FunctionServiceService, {}>,
  functionId: string,
  fileContents: Buffer,
  bucketObjectName: string,
  inputs: ActionInputs,
): Promise<void> {
  core.startGroup('Create function version');
  try {
    core.info(`Function '${inputs.functionName}' ${functionId}`);

    //convert variables
    core.info(`Parsed memory: "${inputs.memory}"`);
    core.info(`Parsed timeout: "${inputs.executionTimeout}"`);

    const request = CreateFunctionVersionRequest.fromJSON({
      functionId,
      runtime: inputs.runtime,
      entrypoint: inputs.entrypoint,
      resources: {
        memory: inputs.memory,
      },
      serviceAccountId: inputs.serviceAccount,
      description: inputs.description,
      environment: parseEnvironmentVariables(inputs.environment),
      executionTimeout: {seconds: inputs.executionTimeout},
    });

    //get from bucket if supplied
    if (inputs.bucket) {
      core.info(`From bucket: "${inputs.bucket}"`);

      request.package = Package.fromJSON({
        bucketName: inputs.bucket,
        objectName: bucketObjectName,
      });
    } else {
      request.content = fileContents;
    }
    // Create new version
    const operation = await functionService.createVersion(request);
    await completion(operation, session);

    handleOperationError(operation);
    core.info('Operation complete');

    const metadata = getMetadata(operation) as CreateFunctionVersionMetadata;
    core.setOutput('version-id', metadata.functionVersionId);
  } finally {
    core.endGroup();
  }
}

async function zipSources(inputs: ActionInputs): Promise<Buffer> {
  core.startGroup('ZipDirectory');

  try {
    const outputStreamBuffer = new streamBuffers.WritableStreamBuffer({
      initialSize: 1000 * KB, // start at 1000 kilobytes.
      incrementAmount: 1000 * KB, // grow by 1000 kilobytes each time buffer overflows.
    });

    const archive = archiver('zip', {zlib: {level: 9}});
    core.info('Archive initialize');

    archive.pipe(outputStreamBuffer);
    const patterns = parseIgnoreGlobPatterns(inputs.excludePattern);
    for (const line of inputs.include) {
      if (fs.lstatSync(line).isDirectory()) {
        archive.directory(line, line, data => {
          const res = !patterns.map(p => minimatch(data.name, p)).some(x => x);
          return res ? data : false;
        });
      } else {
        archive.file(line, {name: line});
      }
      core.info(`Path '${line}' added to archive`);
    }

    await archive.finalize();

    core.info('Archive finalized');

    outputStreamBuffer.end();
    const buffer = outputStreamBuffer.getContents();
    core.info('Buffer object created');

    if (!buffer) {
      throw Error('Failed to initialize Buffer');
    }

    return buffer;
  } finally {
    core.endGroup();
  }
}

function parseIgnoreGlobPatterns(patterns: string[]): string[] {
  const result: string[] = [];

  for (const pattern of patterns) {
    //only not empty patterns
    if (pattern?.length > 0) {
      result.push(pattern);
    }
  }

  core.info(`Source ignore pattern: "${JSON.stringify(result)}"`);
  return result;
}

function parseEnvironmentVariables(env: string[]): {[s: string]: string} {
  core.info(`Environment string: "${env}"`);

  const environment: {[key: string]: string} = {};
  for (const line of env) {
    const [key, ...value] = line.split('=');
    environment[key.trim()] = value.join()?.trim();
  }

  core.info(`EnvObject: "${JSON.stringify(environment)}"`);
  return environment;
}

run();
