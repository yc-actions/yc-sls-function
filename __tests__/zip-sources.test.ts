import { expect, test } from '@jest/globals'
import { parseEnvironmentVariables, parseLockboxVariables, Secret, ZipInputs, zipSources } from '../src/main'
import archiver from 'archiver'

// shows how the runner will run a javascript action with env / stdout protocol
describe('zipSources', function () {
    test('it should add files from include', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['./src'],
            excludePattern: [],
            sourceRoot: '.'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const allStartWithSrc = entries.every(e => e.name.includes('src'))
        expect(allStartWithSrc).toBeTruthy()
    })

    test('it should drop files from if they do not match include patterns', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['./src/*.js'],
            excludePattern: [],
            sourceRoot: '.'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const allStartWithSrc = entries.every(e => e.name.includes('src'))
        expect(allStartWithSrc).toBeTruthy()
        expect(entries.length).toBe(1)
        expect(entries[0].name).toBe('src/func.js')
    })

    test('it should drop files from if they match exclude patterns', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['./src'],
            excludePattern: ['*.txt'],
            sourceRoot: '.'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const allStartWithSrc = entries.every(e => e.name.includes('src'))
        expect(allStartWithSrc).toBeTruthy()
        expect(entries.length).toBe(8)
    })

    test('it should drop folder prefix if sourceRoot provided', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['.'],
            excludePattern: [],
            sourceRoot: './src'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const noneStartWithSrc = entries.every(e => !e.name.includes('src'))

        expect(noneStartWithSrc).toBeTruthy()
        expect(entries.length).toEqual(9)
    })

    test.each([['./src'], ['./src/'], ['src']])(
        'it should respect source root and include only needed files with root %s',
        async sourceRoot => {
            const archive = archiver('zip', { zlib: { level: 9 } })
            const inputs: ZipInputs = {
                include: ['./*.js', 'foo/1.txt'],
                excludePattern: [],
                sourceRoot
            }

            const entries: archiver.EntryData[] = []
            archive.on('entry', e => entries.push(e))
            await zipSources(inputs, archive)

            const noneStartWithSrc = entries.every(e => !e.name.includes('src'))
            expect(noneStartWithSrc).toBeTruthy()
            expect(entries.length).toBe(2)
            entries.sort((a, b) => a.name.localeCompare(b.name))
            expect(entries[0].name).toBe('foo/1.txt')
            expect(entries[1].name).toBe('func.js')
        }
    )

    test('it should add folders', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['./src/foo', './src/bar/*'],
            excludePattern: [],
            sourceRoot: '.'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const allStartWithSrc = entries.every(e => e.name.includes('src'))
        expect(allStartWithSrc).toBeTruthy()
        expect(entries.length).toBe(4)
        expect(entries.map(x => x.name).sort()).toMatchSnapshot()
    })

    test('it should ignore empty lines in include', async () => {
        const archive = archiver('zip', { zlib: { level: 9 } })
        const inputs: ZipInputs = {
            include: ['func.js', 'foo/1.txt', ''],
            excludePattern: [],
            sourceRoot: './src/'
        }

        const entries: archiver.EntryData[] = []
        archive.on('entry', e => entries.push(e))
        await zipSources(inputs, archive)

        const noneStartWithSrc = entries.every(e => !e.name.includes('src'))
        expect(noneStartWithSrc).toBeTruthy()
        expect(entries.length).toEqual(2)
        expect(entries.map(x => x.name).sort()).toMatchSnapshot()
    })
})

describe('lockbox', () => {
    test('it should return right lockbox secrets', () => {
        const input = ['ENV_VAR_1=id/verId/VAR_1', 'ENV_VAR_2=id/verId/VAR_2']
        const result = parseLockboxVariables(input)
        const expected: Secret[] = [
            {
                environmentVariable: 'ENV_VAR_1',
                id: 'id',
                versionId: 'verId',
                key: 'VAR_1'
            },
            {
                environmentVariable: 'ENV_VAR_2',
                id: 'id',
                versionId: 'verId',
                key: 'VAR_2'
            }
        ]
        expect(result).toEqual(expected)
    })

    test.each(['123412343', '123=id', '123=id/verId', '123=id/verId/'])(
        'it should throw error when bad input provided %s',
        input => {
            expect(() => parseLockboxVariables([input])).toThrow()
        }
    )
})

describe('environment', () => {
    test('it should return right lockbox secrets', () => {
        const input = [
            'KEY1=value1',
            'BASE64_KEY1=YmFzZTY0X2VuY29kZWRfd2l0aF9lcXVhbF9zaWduXzE=',
            'BASE64_KEY2=YmFzZTY0X2VuY29kZWRfd2l0aF9lcXVhbF9zaWduXzI==',
            'BASE64_KEY3=YmFzZTY0X2VuY29kZWRfd2l0aF9taWRkbGVfc2lnbl9lcXVhbD1yZXF1aXJlZA==',
            'JUST_KEY_SHOULD_BE_A_KEY=zZTY0X2VuY29kZWRf'
        ]

        const result = parseEnvironmentVariables(input)

        const expected: Record<string, string> = {
            KEY1: 'value1',
            BASE64_KEY1: 'YmFzZTY0X2VuY29kZWRfd2l0aF9lcXVhbF9zaWduXzE=',
            BASE64_KEY2: 'YmFzZTY0X2VuY29kZWRfd2l0aF9lcXVhbF9zaWduXzI==',
            BASE64_KEY3: 'YmFzZTY0X2VuY29kZWRfd2l0aF9taWRkbGVfc2lnbl9lcXVhbD1yZXF1aXJlZA==',
            JUST_KEY_SHOULD_BE_A_KEY: 'zZTY0X2VuY29kZWRf'
        }

        expect(result).toEqual(expected)
    })

    test.each(['123412343', '123=id', '123=id/verId', '123=id/verId/'])(
        'it should throw error when bad input provided %s',
        input => {
            expect(() => parseLockboxVariables([input])).toThrow()
        }
    )
})
