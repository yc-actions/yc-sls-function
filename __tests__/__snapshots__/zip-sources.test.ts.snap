// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zipSources it should add folders 1`] = `
[
  "./src/foo/1.txt",
  "./src/foo/2.txt",
  "src/bar/1.txt",
  "src/bar/2.txt",
]
`;

exports[`zipSources it should ignore empty lines in include 1`] = `
[
  "foo/1.txt",
  "func.js",
]
`;
